sklearn/tree/_splitter.pyx

```diff
@@ -816,20 +960,44 @@ cdef class DensePartitioner:
         SIZE_t best_pos,
         double best_threshold,
         SIZE_t best_feature,
+        SIZE_t best_n_missing,
     ) noexcept nogil:
         """Partition samples for X at the best_threshold and best_feature."""
         cdef:
-            SIZE_t p = self.start
-            SIZE_t partition_end = self.end
+            SIZE_t start = self.start
+            SIZE_t end = self.end
+            SIZE_t p, n_missing, partition_end
             SIZE_t[::1] samples = self.samples
             const DTYPE_t[:, :] X = self.X
-
-        while p < partition_end:
-            if X[samples[p], best_feature] <= best_threshold:
+            SIZE_t end_non_missing = self.end - best_n_missing
+
+        # Move missing values to the end
+        if best_n_missing != 0:
+            n_missing = 0
+            p, partition_end = start, end - 1
+            while p < partition_end and n_missing < best_n_missing:
+                # Finds the lowest partition_end that is not missing
```

Comment:
```suggestion
                # Finds the right-most value that is not missing so that
                # it can be swapped with missing values at its left.
```